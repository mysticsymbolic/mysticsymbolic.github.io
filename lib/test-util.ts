export function withMockConsoleLog(fn: (mock: jest.Mock) => void) {
  const originalLog = console.log;
  const mockLog = jest.fn();
  console.log = mockLog;
  try {
    fn(mockLog);
  } finally {
    console.log = originalLog;
  }
}
