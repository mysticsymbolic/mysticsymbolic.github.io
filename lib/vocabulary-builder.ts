import fs from "fs";
import path from "path";
import cheerio from "cheerio";
import { getSvgBoundingBox } from "./bounding-box";
import { extractSpecs } from "./specs";
import { SvgSymbolData, SvgSymbolElement } from "./svg-symbol";
import toml from "toml";
import { validateSvgSymbolMetadata } from "./svg-symbol-metadata";

const SUPPORTED_SVG_TAG_ARRAY: SvgSymbolElement["tagName"][] = ["g", "path"];
const SUPPORTED_SVG_TAGS = new Set(SUPPORTED_SVG_TAG_ARRAY);

const MY_DIR = __dirname;
const SVG_DIR = path.join(MY_DIR, "..", "svg");
const VOCAB_PATH = path.join(MY_DIR, "_svg-vocabulary.json");
const SVG_EXT = ".svg";

function onlyTags(
  elements: cheerio.Element[] | cheerio.Cheerio
): cheerio.TagElement[] {
  const result: cheerio.TagElement[] = [];

  for (let i = 0; i < elements.length; i++) {
    const el = elements[i];
    if (el.type === "tag") {
      result.push(el);
    }
  }

  return result;
}

function isSupportedSvgTag(
  tagName: string
): tagName is SvgSymbolElement["tagName"] {
  return SUPPORTED_SVG_TAGS.has(tagName as any);
}

const SVG_ATTRIB_TO_PROP_MAP: {
  [key: string]: keyof SvgSymbolElement["props"] | undefined;
} = {
  id: "id",
  fill: "fill",
  stroke: "stroke",
  d: "d",
  "stroke-linejoin": "strokeLinejoin",
  "stroke-linecap": "strokeLinecap",
  "stroke-width": "strokeWidth",
  "fill-rule": "fillRule",
};

function attribsToProps(el: cheerio.TagElement): any {
  const { attribs } = el;
  const result: any = {};

  for (let attrib of Object.keys(attribs)) {
    const prop = SVG_ATTRIB_TO_PROP_MAP[attrib];

    if (!prop) {
      throw new Error(`Unknown SVG attribute '${attrib}' in <${el.tagName}>!`);
    }

    result[prop] = attribs[attrib];
  }

  return result;
}

function serializeSvgSymbolElement(
  $: cheerio.Root,
  el: cheerio.TagElement
): SvgSymbolElement {
  let children = onlyTags(el.children).map((child) =>
    serializeSvgSymbolElement($, child)
  );
  const { tagName } = el;
  if (isSupportedSvgTag(tagName)) {
    return {
      tagName,
      props: attribsToProps(el) as any,
      children,
    };
  }
  throw new Error(`Unsupported SVG element: <${tagName}>`);
}

export function convertSvgMarkupToSymbolData(
  filename: string,
  svgMarkup: string
): SvgSymbolData {
  const name = path.basename(filename, SVG_EXT).toLowerCase();
  const $ = cheerio.load(svgMarkup);
  const svgEl = $("svg");
  const rawLayers = onlyTags(svgEl.children()).map((ch) =>
    serializeSvgSymbolElement($, ch)
  );
  const [specs, layers] = extractSpecs(rawLayers);
  const bbox = getSvgBoundingBox(layers);

  const symbol: SvgSymbolData = {
    name,
    bbox,
    layers,
    specs,
  };
  return symbol;
}

export function build() {
  const filenames = fs.readdirSync(SVG_DIR);
  const vocab: SvgSymbolData[] = [];
  for (let filename of filenames) {
    if (path.extname(filename) === SVG_EXT) {
      let filenames = filename;
      let metaToml: string | null = null;
      const metaFilename = `${path.basename(filename, SVG_EXT)}.toml`;
      const metaFilepath = path.join(SVG_DIR, metaFilename);
      if (fs.existsSync(metaFilepath)) {
        filenames += ` and ${metaFilename}`;
        metaToml = fs.readFileSync(metaFilepath, {
          encoding: "utf-8",
        });
      }
      console.log(`Adding ${filenames} to vocabulary.`);
      const svgMarkup = fs.readFileSync(path.join(SVG_DIR, filename), {
        encoding: "utf-8",
      });
      const symbol = convertSvgMarkupToSymbolData(filename, svgMarkup);
      if (metaToml) {
        symbol.meta = validateSvgSymbolMetadata(toml.parse(metaToml));
      }
      vocab.push(symbol);
    }
  }

  console.log(`Writing ${VOCAB_PATH}.`);
  fs.writeFileSync(VOCAB_PATH, JSON.stringify(vocab, null, 2));
}
