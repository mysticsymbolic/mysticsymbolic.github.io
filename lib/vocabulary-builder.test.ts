import { withMockConsoleLog } from "./test-util";
import { convertSvgMarkupToSymbolData } from "./vocabulary-builder";

const CIRCLE = `<path fill="#ffffff" fill-rule="evenodd" stroke="#000000" stroke-width="4" stroke-linecap="round" stroke-linejoin="round" d="M 360.000 29.751 C 542.791 29.751 690.249 177.209 690.249 360.000 C 690.249 542.791 542.791 690.249 360.000 690.249 C 177.209 690.249 29.751 542.791 29.751 360.000 C 29.751 177.209 177.209 29.751 360.000 29.751 Z"/>`;

const RADIAL_GRADIENT = [
  `<radialGradient id="blargy" cx="49.97%" cy="76.69%" r="68.79%">`,
  `<stop offset="0.00%" style="stop-color:rgb(0,0,0);stop-opacity:1.00" />`,
  `<stop offset="100.00%" style="stop-color:rgb(255,255,255);stop-opacity:1.00" />`,
  `</radialGradient>`,
].join("");

function arrow(color: string) {
  return `<path fill="${color}" fill-rule="evenodd" stroke="none" d="M 360.000 679.153 C 360.001 679.156 372.114 713.074 372.116 713.077 C 372.114 713.076 360.001 701.805 360.000 701.804 C 359.999 701.805 347.886 713.076 347.884 713.077 C 347.886 713.074 359.999 679.156 360.000 679.153 Z"/>`;
}

describe("convertSvgMarkupToSymbolData()", () => {
  it("works with SVGs that just have a path and no specs", () => {
    expect(
      convertSvgMarkupToSymbolData("blarg.svg", `<svg>${CIRCLE}</svg>`)
    ).toMatchSnapshot();
  });

  it("processses radial gradients", () => {
    const result = convertSvgMarkupToSymbolData(
      "blarg.svg",
      `<svg>${CIRCLE}${RADIAL_GRADIENT}</svg>`
    );
    expect(result.defs).toMatchSnapshot();
  });

  it("ignores empty layers", () => {
    const sd1 = convertSvgMarkupToSymbolData(
      "blarg.svg",
      `<svg>${CIRCLE}</svg>`
    );
    const sd2 = convertSvgMarkupToSymbolData(
      "blarg.svg",
      `<svg>${CIRCLE}<g id="boop"></g></svg>`
    );
    expect(sd1).toEqual(sd2);
  });

  it("processes specs", () => {
    const result = convertSvgMarkupToSymbolData(
      "blarg.svg",
      `<svg>${CIRCLE}<g id="specs">${arrow("#ff0000")}</g></svg>`
    );
    expect(result.specs?.anchor).toHaveLength(1);
    expect(result.specs).toMatchSnapshot();
  });

  it("ignores colors in specs it doesn't understand", () => {
    withMockConsoleLog((mockLog) => {
      const result = convertSvgMarkupToSymbolData(
        "blarg.svg",
        `<svg>${CIRCLE}<g id="specs">${arrow("#f1f1f1")}</g></svg>`
      );
      expect(result.specs).toEqual({});
      expect(mockLog).toHaveBeenCalledWith(
        'Not sure what to do with specs path with fill "#f1f1f1", ignoring it.'
      );
    });
  });
});
