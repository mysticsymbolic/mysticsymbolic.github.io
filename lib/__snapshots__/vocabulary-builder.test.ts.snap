// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertSvgMarkupToSymbolData() processes specs 1`] = `
Object {
  "anchor": Array [
    Object {
      "normal": Object {
        "x": 0,
        "y": -1,
      },
      "point": Object {
        "t": 0,
        "x": 360,
        "y": 701.804,
      },
    },
  ],
}
`;

exports[`convertSvgMarkupToSymbolData() works with SVGs that just have a path and no specs 1`] = `
Object {
  "bbox": Object {
    "x": Object {
      "max": 692.249,
      "min": 27.751,
    },
    "y": Object {
      "max": 692.249,
      "min": 27.751,
    },
  },
  "layers": Array [
    Object {
      "children": Array [],
      "props": Object {
        "d": "M 360.000 29.751 C 542.791 29.751 690.249 177.209 690.249 360.000 C 690.249 542.791 542.791 690.249 360.000 690.249 C 177.209 690.249 29.751 542.791 29.751 360.000 C 29.751 177.209 177.209 29.751 360.000 29.751 Z",
        "fill": "#ffffff",
        "fillRule": "evenodd",
        "stroke": "#000000",
        "strokeLinecap": "round",
        "strokeLinejoin": "round",
        "strokeWidth": "4",
      },
      "tagName": "path",
    },
  ],
  "name": "blarg",
  "specs": undefined,
}
`;
